### DISCUSSION TITLE
Can we disable video preview for large files?

### DISCUSSION THREAD

**Discussion by kulkarni005** [directus/directus#24921] [Feature Requests]
### Describe the Bug

I am working on an OTT platform where I upload full-resolution 4K raw files, each with a length of 3-4 hours and approximately 2-5 GB in size. using DigitalOcean Spaces - S3 for storage.

It seems that whenever I try to update anything with a file on the file view page, it starts playing the video. Since the video file is very large and is transferred through the https://mydirectus.com/assets path, it causes the site to go down due to rapid RAM and CPU usage. The site remains stuck until the complete file is downloaded in preview.

Note :  using Cloudflare stream for further compression and transcoding. 

### To Reproduce

Whenever I try to update anything with a file on the file view page, it starts playing the video. Since the video file is very large and is transferred through the https://mydirectus.com/assets path, it causes the site to go down due to rapid RAM and CPU usage. The site remains stuck until the complete file is downloaded in preview.

### Directus Version

10.6.4

### Hosting Strategy

Self-Hosted (Custom)

**Comment by br41nslug** [comment]
It shouldnt be loading the complete file 🤔 Most formats can be streamed by browsers however if the format isnt compatible (or the transcoding/compression interferes) then browsers will fall back to loading the entire file indeed. 

We'll need a way to disable embedded players for both `video` or `audio` as these files can be big and not optimized for web.

Absolutely unrelated to the issue: but i doubt these are RAW video files as with those you'd be talking 100's of gigabytes for 4k

**Comment by kulkarni005** [comment]
> It shouldnt be loading the complete file 🤔 Most formats can be streamed by browsers however if the format isnt compatible (or the transcoding/compression interferes) then browsers will fall back to loading the entire file indeed.
> 
> We'll need a way to disable embedded players for both `video` or `audio` as these files can be big and not optimized for web.
> 
> Absolutely unrelated to the issue: but i doubt these are RAW video files as with those you'd be talking 100's of gigabytes for 4k

Sorry for the lack of clarity in the question. I am referring to the high RAM/CPU usage on the Directus instance, not on the client's browser. The increased resource utilization results in the Directus app on the browser becoming unresponsive, especially during data retrieval.

Mix of RAW + Non-Compressed files, the issue occurs for files larger than 2 GB in size.
Currently about 2 Terrabytes total -  averaging 2-15GB/file.

**Comment by br41nslug** [comment]
> Sorry for the lack of clarity in the question. I am referring to the high RAM/CPU usage on the Directus instance

I understood but it doesnt change the solution 😄 The RAM/CPU usage is purely reading the file from storage (we're not doing any transformations or such on video's) not much we can optimize there but if we never request the file from the App then the API doesnt have to do anything.

**Comment by kulkarni005** [comment]
Honestly, i am confused.

My Storage Adapter is S3.
Uploading file transfer is happening via directus instance.
and downloading/accessing file is also from directus instance.

is there any load on directus on this, can we bypass directus instance and connect s3 public url directly for preview.
Also can directus upload as chunks for large files to s3?

on frontend i am using bucket urls for access.
Thanks for response , and sorry for bothering on xmas weekend eve.

**Comment by domdus** [comment]
@kulkarni005 is it mp4 video files you're using?

I experienced the same CPU/RAM usage with big mp4 files on the instance. It's due to the nature of HTML 5 video loading even bigger chunks on bigger files.
When you continue seeking in the video it will download even more chunks (like 300 MB) simultaneously till complete and eventually collapse on the client. (endless buffering & app freeze)
It's not directus involved in it, I could reproduce it with a simple HTML page on the server and video tag. 

One thing that helps (a bit) is moving the videos info header to the start. (re-encode the videos)
Check the answer here:
https://stackoverflow.com/questions/21686191/can-ffmpeg-place-mp4-metainfo-at-the-beginning-of-the-file

--------------------------

Another option that worked incredibly good was this well explained node/express example that sends custom range headers (like only 1 MB chunks) to the client:
https://blog.logrocket.com/build-video-streaming-server-node/

This eliminated the server load almost completely, I can only recommend you to give it a test and play around with the chunk sizes.

I just don't know how this could be turned into a middleware or something that wraps the directus assets endpoint or provided as an extension.

Unless a project requires full blown HLS or DASH encoding anyway, @br41nslug maybe these custom range response headers is something to consider for advanced core features (extending video capabilities), although I understand it's very specific and therefor never created a feature discussion about it.

Merry Christmas guys 🎄

**Comment by kulkarni005** [comment]
> @kulkarni005 is it mp4 video files you're using?
> 
> I experienced the same CPU/RAM usage with big mp4 files on the instance. It's due to the nature of HTML 5 video loading even bigger chunks on bigger files.
> When you continue seeking in the video it will download even more chunks simultaneously till complete and eventually collapse on the client. (endless buffering & app freeze)
> It's not directus involved in it, I could reproduce it with a simple HTML page on the server and video tag. 
> 
> One thing that helps (a bit) is moving the videos info header to the start. (re-encode the videos)
> Check the answer here:
> https://stackoverflow.com/questions/21686191/can-ffmpeg-place-mp4-metainfo-at-the-beginning-of-the-file
> 
> --------------------------
> 
> Another option that worked incredibly good was this well explained node/express example that sends custom range headers (like only 1 MB chunks) to the client:
> https://blog.logrocket.com/build-video-streaming-server-node/
> 
> This eliminated the server load almost completely, I can only recommend you to give it a test and play around with the chunk sizes.
> 
> I just don't know how this could be turned into a middleware or something that wraps the directus assets endpoint or provided as an extension.
> 
> Unless a project requires full blown HLS encoding anyway, @br41nslug maybe these custom range response headers is something to consider for advanced core features (extending video capabilities), although I understand it's very specific and therefor never created a feature discussion about it.
> 
> Merry Christmas guys 🎄 

Merry Christmas to you too !

I tried loading videos directly with s3 bucket urls , it works flawlessly.

I think issue might somewhere between nginx and directus proxying assets through S3. 

I am using cloudflare stream service for end users streaming.

I m having issues in managing videos within directus. 
Setup on digitalocean droplet - with Plesk control panel + nodejs extension - handled by fusion passenger.

**Comment by domdus** [comment]
Oh indeed I missed the part you're using the cloudflare stream service, not local files. 

Regarding proxying of NGINX in front of directus, I just remember another issue I reported - maybe this can help you, but again this deals with ~local files (VLAN remote location):

https://github.com/directus/directus/issues/15769#issuecomment-1263507183
